gallery_dl/extractor/instagram.py

```diff
@@ -0,0 +1,189 @@
+                yield Message.Url, \
+                    'ytdl:{}/p/{}/'.format(self.root, data['shortcode']), data
+
+    def _extract_shared_data(self, page):
+        return json.loads(text.extract(page,
+                          'window._sharedData = ', ';</script>')[0])
+
+    def _extract_postpage(self, url):
+        page = self.request(url).text
+        shared_data = self._extract_shared_data(page)
+        media = shared_data['entry_data']['PostPage'][0]['graphql']['shortcode_media']
+
+        return {
+            'media_id': media['id'],
+            'shortcode': media['shortcode'],
+            'typename': media['__typename'],
+            'display_url': media['display_url'],
+            'height': int(media['dimensions']['height']),
+            'width': int(media['dimensions']['width']),
+            'comments': int(media['edge_media_to_comment']['count']),
+            'likes': int(media['edge_media_preview_like']['count']),
+            'owner_id': media['owner']['id'],
+            'username': media['owner']['username'],
+            'fullname': media['owner']['full_name'],
+        }
+
+    def _extract_profilepage(self, url):
+        page = self.request(url).text
+        shared_data = self._extract_shared_data(page)
+
+        while True:
+            # Deal with different structure of profile pages: the first page
+            # has interesting data in `entry_data', next pages in `data'.
+            if 'entry_data' in shared_data:
+                base_shared_data = shared_data['entry_data']['ProfilePage'][0]['graphql']
+
+                # `rhx_gis' and `user_id' are available only in the first page
+                rhx_gis = shared_data['rhx_gis']
+                user_id = base_shared_data['user']['id']
+            else:
+                base_shared_data = shared_data['data']
+
+            timeline = base_shared_data['user']['edge_owner_to_timeline_media']
+            has_next_page = timeline['page_info']['has_next_page']
+            shortcodes = [n['node']['shortcode'] for n in timeline['edges']]
+
+            for s in shortcodes:
+                url = '{}/p/{}/'.format(self.root, s)
+                yield self._extract_postpage(url)
+
+            if not has_next_page:
+                break
+
+            end_cursor = timeline['page_info']['end_cursor']
+            variables = '{{"id":"{}","first":12,"after":"{}"}}'.format(
+                user_id,
+                end_cursor,
+            )
+            xigis = '{}:{}'.format(rhx_gis, variables)
+            headers = {
+                "X-Requested-With": "XMLHttpRequest",
+                "X-Instagram-GIS": hashlib.md5(xigis.encode()).hexdigest(),
+            }
+            url = '{}/graphql/query/?query_hash={}&variables={}'.format(
+                self.root,
+                '66eb9403e44cc12e5b5ecda48b667d41',
+                variables,
+            )
+            shared_data = json.loads(self.request(url, headers=headers).text)
```

Comment:
The `response` object returned by `self.request()` has a `.json()` method. Use that instead of "manually" loading the JSON result. -> `shared_data = self.request(...).json()`

Reply:
Whoops, I have adjusted it as suggested!

Reply:
I have converted it to a tuple.